### PR TITLE
New addon r mapcalc.tiled

### PR DESCRIPTION
**Pull request by mlennert**
This pull request proposes a new addon module which cuts a raster input map into tiles and runs r.mapcalc over these tiles before patching the result together into a single output raster map. It is just a simple wrapper over GridModule.

### CODE REVIEW

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+#
+#		This program is free software under the GNU General Public
+#		License (>=v2). Read the file COPYING that comes with GRASS
+#		for details.
+#############################################################################
+
+#%Module
+#% description: Runs r.mapcalc in parallel over tiles
+#% keyword: raster
+#% keyword: mapcalc
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% keyword: algebra
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+#               to allow parallel processing
+# COPYRIGHT:	(C) 1997-2018 by the GRASS Development Team
+#
+#		This program is free software under the GNU General Public
+#		License (>=v2). Read the file COPYING that comes with GRASS
+#		for details.
+#############################################################################
+
+#%Module
+#% description: Runs r.mapcalc in parallel over tiles
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% description: Runs r.mapcalc in parallel over tiles.
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+#%end
+#
+#%option G_OPT_R_OUTPUT
+#% description: Name of raster output map resulting from expression
+#% key: output
+#% required : no
+#%end
+#
+#%option
+#% key: tile_width
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% key: width
```
Reason: parameter compatibility with r.tile

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+#% type: integer
+#% description: Width of tiles
+#% answer: 1000
+#% required: yes
+#%end
+#
+#%option
+#% key: tile_height
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% key: height
```
Reason: parameter compatibility with r.tile

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+#%option G_OPT_R_OUTPUT
+#% description: Name of raster output map resulting from expression
+#% key: output
+#% required : no
+#%end
+#
+#%option
+#% key: tile_width
+#% type: integer
+#% description: Width of tiles
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% description: Width of tiles (columns)
```
Reason: parameter compatibility with r.tile

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+#% answer: 1000
+#% required: yes
+#%end
+#
+#%option
+#% key: tile_height
+#% type: integer
+#% description: Height of tiles
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% description: Height of tiles (rows)
```
Reason: parameter compatibility with r.tile

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+#% answer: 1000
+#% required: yes
+#%end
+#
+#%option
+#% key: overlap
+#% type: integer
+#% description: Overlap between tiles
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% description: Overlap of tiles
```
Reason: parameter description as in r.tile (i18N compatibility)

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+#% answer: 0
+#% required: yes
+#%end
+#
+#%option
+#% key: processes
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% key: nprocs
```
Reason: parameter compatibility with t.rast.mapcalc

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+#% type: integer
+#% description: Number of parallel processes
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
#% description: Number of r.mapcalc processes to run in parallel
```
Reason: parameter description as in t.rast.mapcalc (i18N compatibility)

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+<p>
+The user provides the map calculation <b>expression</b>. The output mapname
+is automatically extracted from that expression by extracting the string before 
+the '='. If the <b>expression</b> is more complex, the user can also provide the
+the output map name with the parameter <b>output</b>
+
+<p>
+Tiles can be defined with the <b>tile_width</b>, <b>tile_height</b> and 
+<b>overlap</b> parameters. If <b>processes</b> is higher than one, these tiles
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
<b>overlap</b> parameters. If <b>nprocs</b> is higher than one, these tiles 
```
(see .py code)

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+<h2>NOTES</h2>
+
+<h2>EXAMPLE</h2>
+
+Run r.mapcalc over tiles with size 1000x1000 using 4 parallel processes:
+
+<div class="code"><pre> 
+g.region rast=ortho_2001_t792_1m
+r.mapcalc.tiled expression = "bright_pixels = if(ortho_2001_t792_1m > 200, 1, 0)" \
+   tile_width=1000 tile_height=1000 processes=4
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
   tile_width=1000 tile_height=1000 nprocs=4
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                   self.start_row, self.start_col, self.out_prefix)
+
+
+def main():
+
+    expression = options['expression']
+    width = int(options['tile_width'])
+    height = int(options['tile_height'])
+    overlap = int(options['overlap'])
+    processes = int(options['processes'])
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
    processes = int(options['nprocs'])
```

**Comment by neteler** [comment]
@mlennert note that this commit contains also "v.class.mlR" code which should be removed from this PR

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+result together into a single output raster map.
+
+<p>
+The user provides the map calculation <b>expression</b>. The output mapname
+is automatically extracted from that expression by extracting the string before 
+the '='. If the <b>expression</b> is more complex, the user can also provide the
+the output map name with the parameter <b>output</b>
+
+<p>
+Tiles can be defined with the <b>tile_width</b>, <b>tile_height</b> and
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
Tiles can be defined with the <b>width</b>, <b>height</b> and 
```
See related change in the .py file

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+The <b>mapset_prefix</b> parameter allows to make sure that the temporary
+mapsets created during the tiled processing have unique names. This is useful
+if the user runs <em>r.mapcalc.tiled</em> several times in parallel (e.g. in
+an HPC environment).
+
+<h2>NOTES</h2>
+
+<h2>EXAMPLE</h2>
+
+Run r.mapcalc over tiles with size 1000x1000 using 4 parallel processes:
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
Run <b>r.mapcalc</b> over tiles with size 1000x1000 using 4 parallel processes
(North Carolina sample dataset):
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+<div class="code"><pre> 
+g.region rast=ortho_2001_t792_1m
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
g.region raster=ortho_2001_t792_1m
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+r.mapcalc.tiled expression = "bright_pixels = if(ortho_2001_t792_1m > 200, 1, 0)" \
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
r.mapcalc.tiled expression="bright_pixels = if(ortho_2001_t792_1m > 200, 1, 0)" \
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python
+
+############################################################################
+#
+# MODULE:	r.mapcalc.tiled
+# AUTHOR(S):	Moritz Lennert
+#
+# PURPOSE:	Run r.mapcalc over tiles of the input map
+#               to allow parallel processing
+# COPYRIGHT:	(C) 1997-2018 by the GRASS Development Team
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
# COPYRIGHT:	(C) 2019 by the GRASS Development Team
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+        rpatch_map(output_map,
+                   self.mset.name, self.msetstr, bboxes,
+                   self.module.flags.overwrite,
+                   self.start_row, self.start_col, self.out_prefix)
+
+
+def main():
+
+    expression = options['expression']
+    width = int(options['tile_width'])
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
    width = int(options['width'])
```
Changed for parameter compatibility with r.tile

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+    height = int(options['tile_height'])
```
Comment (neteler) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]:
```suggestion
    height = int(options['height'])
```
Changed for parameter compatibility with r.tile

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py (hunk 8 above)
Comment (mlennert) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]: Thanks for all these corrections !

On this one, I personally prefer processes which is more in line with the general idea of spelling out parameters and avoiding abbreviations and it is the version we use in a lot of the other addon modules using parallel processing. I guess this means we should make this into a more fundamental discussion on this parameter name.

Comment (neteler) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]: I generally agree, but we have this much (plus testsuite) using the standard `nprocs` parameter - IMHO a task for GRASS GIS 8.0 to change to a better parameter name:

```
ag -l nprocs | grep -v testsuite
vector/v.surf.rst/main.c
lib/python/pygrass/modules/interface/module.py
lib/python/temporal/mapcalc.py
lib/python/temporal/temporal_raster3d_algebra.py
lib/python/temporal/aggregation.py
lib/python/temporal/temporal_raster_algebra.py
lib/python/temporal/extract.py
lib/python/temporal/temporal_raster_base_algebra.py
lib/python/temporal/temporal_algebra.py
lib/cdhc/doc/goodness.ps
gui/wxpython/animation/controller.py
gui/wxpython/datacatalog/tree.py
gui/wxpython/animation/provider.py
gui/wxpython/animation/dialogs.py
gui/wxpython/core/settings.py
temporal/t.rast.to.vect/t.rast.to.vect.html
temporal/t.rast.to.vect/t.rast.to.vect.py
temporal/t.rast.neighbors/t.rast.neighbors.html
temporal/t.rast.neighbors/t.rast.neighbors.py
temporal/t.rast.neighbors/test.t.rast.neighbors.sh
temporal/t.rast3d.extract/t.rast3d.extract.py
temporal/t.rast.mapcalc/t.rast.mapcalc.html
temporal/t.rast.mapcalc/test.t.rast.mapcalc.operators.sh
temporal/t.rast.mapcalc/t.rast.mapcalc.py
temporal/t.rast.mapcalc/test.t.rast.mapcalc.sh
temporal/t.rast.gapfill/t.rast.gapfill.py
temporal/t.rast3d.algebra/t.rast3d.algebra.py
temporal/t.rast.aggregate/t.rast.aggregate.py
temporal/t.rast.aggregate/t.rast.aggregate.html
temporal/t.rast.algebra/t.rast.algebra.py
temporal/t.rast.what/t.rast.what.py
temporal/t.rast.aggregate.ds/t.rast.aggregate.ds.py
temporal/t.rast.contour/t.rast.contour.py
temporal/t.rast.contour/t.rast.contour.html
temporal/t.rast3d.mapcalc/t.rast3d.mapcalc.py
temporal/t.rast3d.mapcalc/test.t.rast3d.mapcalc.sh
temporal/t.rast.extract/t.rast.extract.py
temporal/t.vect.extract/test.t.vect.extract.sh
temporal/t.vect.extract/t.vect.extract.py
raster/r.sun/main.c
raster/r.sim/r.sim.sediment/main.c
raster/r.sim/r.sim.water/main.c
raster3d/r3.gradient/main.c
```

**Comment by neteler** [comment]
> > @mlennert note that this commit contains also "v.class.mlR" code which should be removed from this PR
> 
> Oops. Still learning git. I think I created the branch for this PR from the branch of my previous PR, and not from master...
> Now I have to find out how to undo this...

It happened to me as well and I found this nice answer here which does the trick:

https://stackoverflow.com/questions/12481639/remove-files-from-git-commit

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py (hunk 8 above)
Comment (neteler) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]: ... of course no showstopper for me! Happy to see this new addon available

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+<em>r.mapcalc.tiled</em> cuts a raster input map into tiles and runs 
+<a href="r.mapcalc.html">r.mapcalc</a> over these tiles before patching the
+result together into a single output raster map.
+
+<p>
+The user provides the map calculation <b>expression</b>. The output mapname
+is automatically extracted from that expression by extracting the string before 
+the '='. If the <b>expression</b> is more complex, the user can also provide the
+the output map name with the parameter <b>output</b>
```
Comment (veroandreo) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
the output mapname with the parameter <b>output</b>
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+
+<p>
+Tiles can be defined with the <b>width</b>, <b>height</b> and 
+<b>overlap</b> parameters. If <b>processes</b> is higher than one, these tiles 
+will be processed in parallel.
+
+<p>
+The <b>mapset_prefix</b> parameter allows to make sure that the temporary
```
Comment (veroandreo) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
The <b>mapset_prefix</b> parameter ensures that the temporary
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+mapsets created during the tiled processing have unique names. This is useful
```
Comment (veroandreo) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]:
```suggestion
mapsets created during the tiles processing have unique names. This is useful
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+
+<h2>SEE ALSO</h2>
+
+<a href="r.mapcalc.html">r.mapcalc</a> 
+
+<h2>AUTHOR</h2> 
+
+Moritz Lennert
+
+<p><i>Last changed: $Date: 2019-04-19 08:30:49 +0200 (ven 19 avr 2019) $</i>
```
Comment (veroandreo) [inline grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]: 08:30:49 +0200 (ven 19 avr 2019) <-- is all this needed?

Comment (neteler) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]: It should really be like this:

```suggestion
<p><i>Last changed: $Date$</i>
```

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html (hunk 20 above)
Comment (mlennert) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]: map name looks better to me...

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html (hunk 22 above)
Comment (mlennert) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.html]: It is about the processing which is tiled, not the processing of the tiles...

File: grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py (hunk 8 above)
Comment (mlennert) [reply grass7/raster/r.mapcalc.tiled/r.mapcalc.tiled.py]: Leaving as "processes" for now. It we decide to harmonize, I should change it also in all our other modules (i.segment.uspo, i.segment.stats, etc).

**Pull request closed without merging**